File: src/app/calibration/series/ts-linear.series.spec.ts
import { testCases } from "./regression-test-data.spec";
import { TSLinearSeries } from "./ts-linear.series";

fdescribe("Theil Sen Linear Regression", (): void => {
    const testMaxSize = 7;
    let tsReg: TSLinearSeries;

    beforeEach((): void => {
        tsReg = new TSLinearSeries(testMaxSize);
        testCases.forEach((testCase: Array<number>): void => {
            tsReg.push(testCase[1] / 1e6, testCase[0] / 1e6);
        });
    });

    it("should correctly calculate Median", (): void => {
        const expectedMedian = -39.37665713543009;
        expect(tsReg.median()).toEqual(expectedMedian);
    });

    it("should assign the median to coefficientA", (): void => {
        expect(tsReg.median()).toEqual(tsReg.coefficientA());
    });

    it("should calculate coefficientB correctly", (): void => {
        const tsRegCoeffB: TSLinearSeries = new TSLinearSeries(testMaxSize);

        testCases.forEach((testCase: Array<number>): void => {
            tsRegCoeffB.push(testCase[1] / 1e6, testCase[0] / 1e6);
            expect(tsRegCoeffB.coefficientB()).toEqual(testCase[3]);
        });
    });

    it("should be empty after reset", (): void => {
        tsReg.reset();
        expect(tsReg.median()).toEqual(0);
    });
});

File: test/unit/series/ts-linear-series.spec.cpp
#include "catch2/catch_test_macros.hpp"

#include "../../../src/utils/series/ts-linear-series.h"
#include "./regression.test-cases.spec.h"

TEST_CASE("Theil Sen Linear Regression", "[regression]")
{
    const auto testMaxSize = 7U;
    TSLinearSeries tsReg(testMaxSize);

    for (const auto &testCase : testCases)
    {
        tsReg.push(testCase[1] / 1e6, testCase[0] / 1e6);
    }

    SECTION("should correctly calculate Median")
    {
        const auto expectedMedian = -39.37665713543009;
        REQUIRE(tsReg.median() == expectedMedian);
    }

    SECTION("should assign the median to coefficientA")
    {
        REQUIRE(tsReg.median() == tsReg.coefficientA());
    }

    SECTION("should calculate coefficientB correctly")
    {
        TSLinearSeries tsRegCoeffB(testMaxSize);

        for (const auto &testCase : testCases)
        {
            tsRegCoeffB.push(testCase[1] / 1e6, testCase[0] / 1e6);
            REQUIRE(tsRegCoeffB.coefficientB() == testCase[3]);
        }
    }

    SECTION("should be empty after reset")
    {
        tsReg.reset();
        REQUIRE(tsReg.median() == 0);
    }
}